### PR TITLE
Downgrade zvariant_derive to make MSRV 1.46 again

### PR DESCRIPTION
**Pull request by palfrey**
https://github.com/palfrey/notify-rust/actions/runs/1273105903 demos this working, as this would need #117 merged to build here.

### CODE REVIEW

**Comment by hoodie** [comment]
thanks - but is there a good reason to stick to the old version?

**Comment by palfrey** [comment]
> thanks - but is there a good reason to stick to the old version?

Mainly to avoid compiler upgrades in other things using this. I originally hit issues with this with installing `cargo-watch`, as my actual app used 1.47.0. You can just upgrade the version, but I'm generally of the theory that libraries should use the lowest Rust version that doesn't get in the way of them using any features they need i.e. if `notify-rust` used any features of the later `zvariant_derive` or there was a security issue that was only fixed in later versions that's one thing, but as it is AFAIK the older versions work fine and they permit a lower Rust version.

**Comment by hoodie** [comment]
I get your point, but I don't think we should do this here. the version description of zvariant_derive in notify-rust is already being explicitly vague. It's not like we're checking in a lock file right, it just says `"2"`.
So by enforcing ">=2, <2.7" we're actually making it more strict. What if there's an import fix in zvariant_derive 2.8?
I don't think a lib should enforce this, as you said, so maybe it is up to the consumer to force it to use a specific version of zvariant_derive. notify-rust should be working fine with either one. My vote for MSRV is only a recommendation and the fact that it uses `>=2.7` in CI is due to the missing `Cargo.lock`.

**Comment by palfrey** [comment]
Yeah, that's a good point. `notify-rust` should specify what it's happy with, which is any version of `zvariant_derive` at the moment. It would be nice to check against 1.46.0 with just that build limiting it to a lower version, which is fiddly at the moment. I tried doing something with [`minimal-versions`](https://doc.rust-lang.org/cargo/reference/unstable.html#minimal-versions) and a nightly build around the 1.46.0 release, but ran into issues with zvariant not having the right dependency values on `serde_repr`. I'm going to file an issue around that, and see what I can do there, and maybe revisit after that's gotten fixed.

Closing for the moment, as this probably isn't the right way to do it anyways. Thanks for your time!

**Comment by hoodie** [comment]
I think it's ok that sometimes stuff breaks and you're forced to update. In other languages not breaking things this way has lead to people building build systems that make it extremely hard to update. at least with rust you can just rebuild and you're good. Or are you trying to run on something like Debian stable?

**Comment by palfrey** [comment]
> I think it's ok that sometimes stuff breaks and you're forced to update. In other languages not breaking things this way has lead to people building build systems that make it extremely hard to update. at least with rust you can just rebuild and you're good. Or are you trying to run on something like Debian stable?

I am actually :) But I've got `rustup` installed, so version upgrades are pretty trivial. I'm not quite so masochistic as to want to use the Debian packaged version of rust! 

The thing that bugs me is that things like the [resolver issue](https://github.com/hoodie/notify-rust/issues/114) are often not very clear that the true issue is "you don't have a high enough Rust version", and so if a library doesn't need a higher Rust version, I'd prefer it didn't do things that make it require one. Also, and this is the case here, many of these issues are to do with some specific library further down in the dependency tree and given the lack of proper MSRV-based dependency resolution (e.g. if I'm running 1.46, do the `zvariant_derive` thing here, but install a later version if I'm using a newer Rust) Cargo can't do the right thing here.

TL;DR: this a problem that bugs me, and so I try and contribute fixes to the problem to libraries when I find it.
